### PR TITLE
Support parsing nested JSON properties and schema check for nested properties

### PR DESCRIPTION
**Pull request by linyguo**
- Add a function to support parsing nested properties
- Update test scenarios for nested properties and schema check

### CODE REVIEW

File: api/pkg/apis/v1alpha1/utils/utils.go [hunk 1]
```diff
@@ -321,6 +322,35 @@ func jsonPathQuery(obj interface{}, jsonPath string) (interface{}, error) {
...
+		if _, ok := result.(error); ok {
+			return nil, false
+		}
+		value = result
+	}
+	return value, value != nil
+}
+
+func formatPathForNestedJsonField(s string) string {
+	if len(s) == 0 || (len(s) > 0 && s[0] == '.') {
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/utils/utils.go:348]: You can just write len(s)==0 || s[0] == '.'

File: api/pkg/apis/v1alpha1/utils/utils_test.go [hunk 2]
```diff
@@ -404,3 +404,85 @@ func TestToInterfaceMap(t *testing.T) {
...
+	m := map[string]interface{}{
+		"a": map[string]interface{}{
+			"b": map[string]interface{}{
+				"c": value,
+			},
+		},
+	}
+	m2, ok := JsonParseProperty(m, "a..b.c")
+	assert.False(t, ok)
+	assert.Equal(t, m2, nil)
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/utils/utils_test.go:487]: Consider adding a test case in which evaluation stop in the middle of the whole path?

File: api/pkg/apis/v1alpha1/utils/utils_test.go [hunk 3]
```diff
@@ -404,3 +404,85 @@ func TestToInterfaceMap(t *testing.T) {
...
+		"a": map[string]interface{}{
+			"b": map[string]interface{}{
+				"c": value,
+			},
+		},
+	}
+	m2, ok := JsonParseProperty(m, "a.b.c")
+	assert.True(t, ok)
+	assert.Equal(t, value, m2)
+}
```
Comment (lirenjie95) [inline api/pkg/apis/v1alpha1/utils/utils_test.go:446]: I'm curious if there's any field can contain "." like
```go
m := map[string]interface{}{
		"a": map[string]interface{}{
			"b.c": map[string]interface{}{
				"c": value,
			},
		},
	}
```

File: api/pkg/apis/v1alpha1/utils/utils_test.go (hunk 2 above)
Comment (Haishi2016) [reply api/pkg/apis/v1alpha1/utils/utils_test.go:487]: Agreed

File: api/pkg/apis/v1alpha1/utils/schema_test.go [hunk 4]
```diff
@@ -281,6 +332,23 @@ func TestExpression(t *testing.T) {
...
+	}
+	properties := map[string]interface{}{
+		"person": map[string]interface{}{
+			"expression": "13",
+		},
+	}
+	result, err := schema.CheckProperties(properties, nil)
+	assert.Nil(t, err)
+	assert.True(t, result.Valid)
+}
```
Comment (iwangjintian) [inline api/pkg/apis/v1alpha1/utils/schema_test.go:351]: We can have some error case for the 4 rule types

File: api/pkg/apis/v1alpha1/utils/utils.go (hunk 1 above)
Comment (linyguo) [reply api/pkg/apis/v1alpha1/utils/utils.go]: removed this line to adopt jq syntax.

File: api/pkg/apis/v1alpha1/utils/utils.go [hunk 5]
```diff
@@ -323,6 +324,35 @@ func jsonPathQuery(obj interface{}, jsonPath string) (interface{}, error) {
...
+		}
+		if _, ok := result.(error); ok {
+			return nil, false
+		}
+		value = result
+	}
+	return value, value != nil
+}
+
+func formatPathForNestedJsonField(s string) string {
```
[the remark this answers is not in the document]
Comment (linyguo) [reply api/pkg/apis/v1alpha1/utils/utils.go:351]: added test case for filter and array in jq query. Will test more scenarios later on.

File: api/pkg/apis/v1alpha1/utils/utils_test.go (hunk 3 above)
Comment (linyguo) [reply api/pkg/apis/v1alpha1/utils/utils_test.go]: Updated to allow the dot name scenario

File: api/pkg/apis/v1alpha1/providers/config/catalog/catalogprovider_test.go [hunk 6]
```diff
@@ -95,7 +101,7 @@ func TestRead(t *testing.T) {
 	}
 	assert.Nil(t, err)
 
-	res, err := provider.Read("catalog1:v1", "components", nil)
+	res, err := provider.Read("catalog1:v1", ".components", nil)
```
Comment (Haishi2016) [inline api/pkg/apis/v1alpha1/providers/config/catalog/catalogprovider_test.go:104]: So, we always require the "." now? I think we should make it optional.

Comment (Haishi2016) [reply api/pkg/apis/v1alpha1/providers/config/catalog/catalogprovider_test.go:104]: I think we should still make simple cases straight-forward - like querying a property directly. Can we check if it's a valid json identifier first?

Comment (linyguo) [reply api/pkg/apis/v1alpha1/providers/config/catalog/catalogprovider_test.go]: updated. Add an input check for simple query strings.

File: api/pkg/apis/v1alpha1/utils/utils.go [hunk 7]
```diff
@@ -323,6 +324,51 @@ func jsonPathQuery(obj interface{}, jsonPath string) (interface{}, error) {
 	}
 }
 
+func isAlphanum(query string) bool {
+	return regexp.MustCompile(`^[a-zA-Z0-9]+$`).MatchString(query)
```
Comment (Haishi2016) [inline api/pkg/apis/v1alpha1/utils/utils.go:328]: Can we do something like ^(?:\.|\[|\(|length|keys|flatten|map|select|unique|sort|join|add|reduce)

Comment (linyguo) [reply api/pkg/apis/v1alpha1/utils/utils.go:328]: As discussed offline, update to check if string is quoted by `` to determine if it is a jq expression.

**Review by iwangjintian — APPROVED** [review]

**Review by msftcoderdjw — DISMISSED** [review]

File: api/pkg/apis/v1alpha1/providers/config/catalog/catalogprovider_test.go [hunk 8]
```diff
@@ -107,7 +113,24 @@ func TestRead(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, "name", summary[0].Name)
 
-	res, err = provider.Read(ctx, "catalog1:v1", "parentAttribute", nil)
+	res, err = provider.Read(ctx, "catalog1:v1", "`.a.b.c`", nil)
```
Comment (Haishi2016) [inline api/pkg/apis/v1alpha1/providers/config/catalog/catalogprovider_test.go:116]: Is .a.b.c considered a jq query? Shouldn't it be wrapped in single quotes?

Comment (linyguo) [reply api/pkg/apis/v1alpha1/providers/config/catalog/catalogprovider_test.go:116]: Yes, it is a jq query, it is quoted with ``, which will be recognized by jq as directly

**Review by msftcoderdjw — APPROVED** [review]

**Comment by msftcoderdjw** [comment]
Fixed #210

**Pull request merged**